File: Framework/framework.c
//Framework program in c.

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#define DATAFILE_NAME "framework2.dat"
#define FIELDFILE_NAME "fields.cfg"
#define MENUFILE_NAME "menu.cfg"
void createRecord();
void readRecords();
void updateRecord();
void deleteRecord();
char fieldName[50][100];
int countOfFields;
FILE *fpFields;
void main()
{
	int option;
	char menuLine[100];
	FILE *fpMenu;
	do
	{
		fpFields = fopen(FIELDFILE_NAME, "r");
		int counter = 0;
		system("cls");
		while(fgets(fieldName[counter], 100, fpFields))
		{
			fieldName[counter][strlen(fieldName[counter]) - 1] = '\0';
			counter++;
		}
		countOfFields = counter;
		fpMenu = fopen(MENUFILE_NAME, "r");
		while(fgets(menuLine, 100, fpMenu))
		{
			printf("%s", menuLine);
		}
		printf("\nEnter your option: ");
		scanf("%d", &option);
		switch(option)
		{
			case 1: createRecord();
			break;
			case 2: readRecords();
			break;
			case 3: updateRecord();
			break;
			case 4: deleteRecord();
			break;
			case 5: ;
			break;
			default: printf("\nEnter a valid option");
			break;
			getch();
		}
	}while(option != 5);
	fclose(fpMenu);
	fclose(fpFields);
}
void createRecord()
{
	printf("\nEnter new record details ...\n");
	char fieldValue[20], status[20];
	int counter = 0;
	FILE *fpData;
	fpData = fopen(DATAFILE_NAME, "a");
	if(fpData == NULL)
	{
		printf("\nError!");
		return;
	}
	strcpy(status, "Activated");
	fwrite(&status, sizeof(status), 1, fpData);
	while(counter < countOfFields)
	{
		printf("Please Enter %s: ", fieldName[counter]);
		scanf("%s", fieldValue);
		fwrite(fieldValue, sizeof(fieldValue), 1, fpData);
		counter++;
	}
	printf("\nRecord saved successfully.\nEnter a key to continue.");
	fclose(fpData);
	getch();
}

void readRecords()
{
	printf("\nALL THE RECORDS\n");
	char fieldValue[20], status[20], status1[20];
	int counter = 0;
	FILE *fpData;
	strcpy(status, "Activated");
	strcpy(status1, "Deactivated");
	fpData = fopen(DATAFILE_NAME, "r");
	if(fpData == NULL)
	{
		printf("\nError!");
		return;
	}
	while(fread(fieldValue, sizeof(fieldValue), 1, fpData))
	{
		if(strcmp(status, fieldValue) == 0 || strcmp(status1, fieldValue) == 0)
		{
			printf("status: ");
			printf("%s\n", fieldValue);
		}
		else
		{
			fgets(fieldName[counter], 100, fpFields);
			printf("%s: %s\n", fieldName[counter], fieldValue);
			counter++;
			if(counter == countOfFields)
			{
				printf("----------------\n");
				counter = 0;
			}
		}
	}
	printf("\nEnter a key to continue.");
	fclose(fpData);
	getch();
}

void updateRecord()
{
	char fieldValue[20], tempFieldValue[20];
	FILE *fpData;
	int counter = 0;
	fpData = fopen(DATAFILE_NAME, "r+");
	if(fpData == NULL)
	{
		printf("\nError!");
		return;
	}
	fgets(fieldName[counter], 20, fpFields);
	printf("\nTo update enter your %s: ", fieldName);
	scanf("%s", tempFieldValue);
	while(fread(fieldValue, sizeof(fieldValue), 1, fpData))
	{
		if(strcmp(tempFieldValue, fieldValue) == 0)
		{
			fseek(fpData, 20, SEEK_CUR);
			fread(fieldValue, sizeof(fieldValue), 1, fpData);
			printf("%s should be changed to: ", fieldValue);
			scanf("%s", fieldValue);
			fseek(fpData, -20, SEEK_CUR);
			fwrite(fieldValue, sizeof(fieldValue), 1, fpData);
			break;
		}
	}
	printf("\nRecord details updated successfully.\nEnter a key to continue.");
	fclose(fpData);
	getch();
}
void deleteRecord()
{
	char fieldValue[20], tempFieldValue[20];
	int counter = 0;
	FILE *fpData;
	fpData = fopen(DATAFILE_NAME, "r+");
	if(fpData == NULL)
	{
		printf("\nError!");
		return;
	}
	fgets(fieldName[counter], 20, fpFields);
	printf("\nTo delete enter your %s: ", fieldName);
	scanf("%s", tempFieldValue);
	while(fread(fieldValue, sizeof(fieldValue), 1, fpData))
	{
		if(strcmp(tempFieldValue, fieldValue) == 0)
		{
			fseek(fpData, -40, SEEK_CUR);
			fread(fieldValue, sizeof(fieldValue), 1, fpData);
			printf("Enter the status as Deactivated: ");
			scanf("%s", fieldValue);
			fseek(fpData, -20, SEEK_CUR);
			fwrite(fieldValue, sizeof(fieldValue), 1, fpData);
			break;
		}
	}
	printf("\nDetails deactivated successfully.\nEnter a key to continue.");
	fclose(fpData);
	getch();
}	

File: Programs/framework.py
# Framework program in python.

dataFile = "framework.dat"
menuFile = "menu.cfg"
fieldsFile = "fields.cfg"
records = {}
fieldValues = []
fData = open(dataFile, 'r').read()
if fData == None:
	print("No data in the file.")
else:
	records = eval(fData)
fFields = open(fieldsFile, "r")
fieldNames = fFields.readlines()
fields = []
fieldCounter = 0
for fieldName in fieldNames:
	if fieldCounter == 0:
		fieldName = fieldName.strip()
		fieldValue = fieldName
		fieldCounter += 1
	else:
		fieldName = fieldName.strip()
		fields.append(fieldName)
def createRecord():
	while(True):
		fieldCounter = 0
		for fieldName in fieldNames:
			if fieldCounter == 0:
				fieldName = fieldName.strip()
				print("Enter " + fieldName + ": ", end = "")
				fieldValue1 = input()
				fieldCounter += 1
			else:
				fieldName = fieldName.strip()
				print("Enter " + fieldName + ": ", end = "")
				fieldValue2 = input()
				fieldValues.append(fieldValue2)
		records[fieldValue1] = fieldValues
		fData = open(dataFile, 'w')
		fData.write(str(records))
		print("Record saved successfully.")
		fData.close()
		break
def displayRecords():
	for record in records:
		counter = 0
		count = 0
		for fieldName in fieldNames:
			if counter == 0:
				fieldName = fieldName.strip()
				print(fieldName + ": " + record)
				counter += 1
			else:
				fieldName = fieldName.strip()
				print(fieldName + ": " + records[record][count])
				count += 1
		print("-----------")
def updateRecord():
	fieldValue1 = input("Enter field value to update the fields: ")
	counter = 1
	for fieldName in fields:
		print(counter, '. ', fieldName, sep = '')
		counter += 1
	option = int(input("Enter your option: "))
	for record in records:
		if record == fieldValue1:
			print("Enter new", fields[option - 1] + ": ", end = '')
			records[record][option - 1] = input()
			fData = open(dataFile, 'w')
			fData.write(str(records))
			print("Record datails are updated successfully.")
			fData.close()
def deleteRecord():
	fieldValue1 = input("Enter field value to delete: ")
	if fieldValue1 in records:
		del records[fieldValue1]
		fData = open(dataFile, 'w')
		fData.write(str(records))
		print("Record deleted successfully.")
		fData.close()
	else:
		print("Record is not found.")
def exit():
	quit()
while(1):
	fmenu = open(menuFile, 'r')
	menuLine = fmenu.read()
	print(menuLine)
	options = [createRecord, displayRecords, updateRecord, deleteRecord, exit]
	length = len(options)
	option = int(input("Enter your option: "))
	if option > length:
		print("Enter a valid option.")
	else:
		options[option - 1]()
	choice = input("Enter any key to continue or enter E to exit: ")
	if choice.upper() == 'E':
		break

